File: 305. Number of Islands II.py
class UF:
    def __init__(self):
        self.ids = {}
        self.count = 0
        self.size = {}
        
    def root(self, A):
        while A != self.ids[A]:
            self.ids[A] = self.ids[self.ids[A]]
            A = self.ids[A]
        return A
    
    def add(self, A):
        self.ids[A] = A
        self.size[A] = 1
        self.count += 1
        
    def union(self, A, B):
        i = self.root(A)
        j = self.root(B)
        if i == j:
            return
        if self.size[i] > self.size[j]:
            i, j = j, i
        self.ids[i] = j
        self.size[j] += self.size[i]
        self.count -= 1

class Solution:
    def numIslands2(self, m: int, n: int, positions: List[List[int]]) -> List[int]:
        '''
        Union find solution
        '''
        uf = UF()
        res = []
        for i,j in positions:
            if (i,j) in uf.ids:
                res.append(uf.count)
                continue
            uf.add((i,j))
            if (i-1,j) in uf.ids:
                uf.union((i,j), (i-1,j))
            if (i+1,j) in uf.ids:
                uf.union((i,j), (i+1,j))
            if (i,j-1) in uf.ids:
                uf.union((i,j), (i,j-1))
            if (i,j+1) in uf.ids:
                uf.union((i,j), (i,j+1))
            res.append(uf.count)
        return res

File: 305. Number of Islands II.java
class Solution {
    public List<Integer> numIslands2(int m, int n, int[][] positions) {
        UF uf = new UF(m*n);
        List<Integer> ans = new ArrayList<>();
        for (int[] pos : positions) {
            int r=pos[0], c=pos[1];
            uf.add(r*n+c);
            if (r > 0 && uf.id[(r-1)*n+c] >= 0) {
                uf.union(r*n+c, (r-1)*n+c);
            }
            if (r < m-1 && uf.id[(r+1)*n+c] >= 0) {
                uf.union(r*n+c, (r+1)*n+c);
            }
            if (c > 0 && uf.id[r*n+c-1] >= 0) {
                uf.union(r*n+c, r*n+c-1);
            }
            if (c < n-1 && uf.id[r*n+c+1] >= 0) {
                uf.union(r*n+c, r*n+c+1);
            }
            ans.add(uf.count);
        }
        return ans;
    }
}
class UF {                          
    public int[] id, size;
    public int count;
    public UF(int n) {
        id = new int[n];
        size = new int[n];
        Arrays.fill(id, -1);
        Arrays.fill(size, 1);
    }
    
    public int root(int A) {
        while (A != id[A]) {
            id[A] = id[id[A]];
            A = id[A];
        }
        return A;
    }
    
    public boolean find(int A, int B) {
        return root(A) == root(B);
    }
    
    public void union(int A, int B) {
        int x = root(A);
        int y = root(B);
        if (x != y) {
            if (size[x] > size[y]) {
                id[y] = x;
                size[x] += size[y];
            } else {
                id[x] = y;
                size[y] += size[x];
            }
            count--;
        }
    }
    
    public void add(int A) {
        if (id[A] < 0) {
            id[A] = A;
            count++;
        }
    }
}